spring-integration-core/src/main/java/org/springframework/integration/config/IntegrationRegistrar.java

```diff
@@ -459,52 +458,11 @@ private void registerListCapableArgumentResolvers(BeanDefinitionRegistry registr
 
 	private BeanDefinitionBuilder internalArgumentResolversBuilder(BeanDefinitionRegistry registry,
 			boolean listCapable) {
-		ManagedList<BeanReference> resolvers = new ManagedList<>();
-		if (!registry.containsBeanDefinition(IntegrationContextUtils.PAYLOAD_RESOLVER_BEAN_NAME)) {
-			BeanDefinitionBuilder resolverBuilder = BeanDefinitionBuilder.genericBeanDefinition(
-					PayloadExpressionArgumentResolver.class);
-			registry.registerBeanDefinition(IntegrationContextUtils.PAYLOAD_RESOLVER_BEAN_NAME,
-					resolverBuilder.getBeanDefinition());
-		}
-		resolvers.add(new RuntimeBeanReference(IntegrationContextUtils.PAYLOAD_RESOLVER_BEAN_NAME));
-
-		if (!registry.containsBeanDefinition(IntegrationContextUtils.PAYLOADS_RESOLVER_BEAN_NAME)) {
-			BeanDefinitionBuilder resolverBuilder = BeanDefinitionBuilder.genericBeanDefinition(
-					PayloadsArgumentResolver.class);
-			registry.registerBeanDefinition(IntegrationContextUtils.PAYLOADS_RESOLVER_BEAN_NAME,
-					resolverBuilder.getBeanDefinition());
-		}
-		resolvers.add(new RuntimeBeanReference(IntegrationContextUtils.PAYLOADS_RESOLVER_BEAN_NAME));
-
-		if (listCapable) {
-			if (!registry.containsBeanDefinition(IntegrationContextUtils.LIST_COLLECTION_RESOLVER_BEAN_NAME)) {
-				BeanDefinitionBuilder resolverBuilder = BeanDefinitionBuilder.genericBeanDefinition(
-						CollectionArgumentResolver.class);
-				resolverBuilder.addConstructorArgValue(listCapable);
-				registry.registerBeanDefinition(IntegrationContextUtils.LIST_COLLECTION_RESOLVER_BEAN_NAME,
-						resolverBuilder.getBeanDefinition());
-			}
-			resolvers.add(new RuntimeBeanReference(IntegrationContextUtils.LIST_COLLECTION_RESOLVER_BEAN_NAME));
-		}
-		else {
-			if (!registry.containsBeanDefinition(IntegrationContextUtils.COLLECTION_RESOLVER_BEAN_NAME)) {
-				BeanDefinitionBuilder resolverBuilder = BeanDefinitionBuilder.genericBeanDefinition(
-						CollectionArgumentResolver.class);
-				resolverBuilder.addConstructorArgValue(listCapable);
-				registry.registerBeanDefinition(IntegrationContextUtils.COLLECTION_RESOLVER_BEAN_NAME,
-						resolverBuilder.getBeanDefinition());
-			}
-			resolvers.add(new RuntimeBeanReference(IntegrationContextUtils.COLLECTION_RESOLVER_BEAN_NAME));
-		}
-
-		if (!registry.containsBeanDefinition(IntegrationContextUtils.MAP_RESOLVER_BEAN_NAME)) {
-			BeanDefinitionBuilder resolverBuilder = BeanDefinitionBuilder.genericBeanDefinition(
-					MapArgumentResolver.class);
-			registry.registerBeanDefinition(IntegrationContextUtils.MAP_RESOLVER_BEAN_NAME,
-					resolverBuilder.getBeanDefinition());
-		}
-		resolvers.add(new RuntimeBeanReference(IntegrationContextUtils.MAP_RESOLVER_BEAN_NAME));
-
+		ManagedList<HandlerMethodArgumentResolver> resolvers = new ManagedList<>();
+		resolvers.add(new PayloadExpressionArgumentResolver());
```

Comment:
Well, unfortunately this doesn't trigger the `BeanFactory` callbacks. Must be like:
```
resolvers.add(new RootBeanDefinition(PayloadExpressionArgumentResolver.class));
```
Will fix on merge if only concern.

Reply:
Thanks.

Reply:
Interesting that all out tests pass, though - I guess we don't have tests that actually need those callbacks 😦

Reply:
I think so. I'll try to come up with something simple during merge. For example something like `#jsonPath()` from the `@Payload(expression)`